### PR TITLE
Feature/update api

### PR DESCRIPTION
**Pull request by pavlo-hilei**
Make UserInput immutable
Add hash to the UserInput and use it for creation of temporary file

### CODE REVIEW

File: lib/core/transpiler_session/transpiler_session.h [hunk 1]
```diff
@@ -45,22 +45,29 @@ struct TranspilerSession {
...
+        _headers = _output.normalized.headers;
     }
 
    private:
     // TODO add methods for user input/output
-    UserInput _input;
+    const UserInput _input;
     UserOutput _output;
 
+    std::string _source;                          ///< Current source code (changes between stages)
```
Comment (IuriiKobein) [inline lib/core/transpiler_session/transpiler_session.h:68]: Please wrap it into structure and give it name - "StagedFile" or ask ChatGPT to help

File: lib/core/transpiler_session/transpiler_session.h [hunk 2]
```diff
@@ -45,22 +45,29 @@ struct TranspilerSession {
...
 
     UserOutput& getOutput() { return _output; }
 
-    void moveOutputToInput() {
-        _input.source = _output.normalized.source;
-        _input.headers = _output.normalized.headers;
+    const std::string& getSource() const { return _source; }
+    std::string& getSource() { return _source; }
+
+    const std::map<std::string, std::string>& getHeaders() const { return _headers; }
```
Comment (IuriiKobein) [inline lib/core/transpiler_session/transpiler_session.h:55]: Add 'Staged' name before -> 'getStagedHeaders' or something else to explain pipeline nature of those files

File: lib/core/transpiler_session/transpiler_session.h [hunk 3]
```diff
@@ -45,22 +45,29 @@ struct TranspilerSession {
...
-    UserInput& getInput() { return _input; }
-
     const UserOutput& getOutput() const { return _output; }
 
     UserOutput& getOutput() { return _output; }
 
-    void moveOutputToInput() {
-        _input.source = _output.normalized.source;
-        _input.headers = _output.normalized.headers;
+    const std::string& getSource() const { return _source; }
```
Comment (IuriiKobein) [inline lib/core/transpiler_session/transpiler_session.h:52]: Add 'Staged' name before -> 'getStagedSource' or something else to explain pipeline nature of this file

File: lib/pipeline/stages/transpiler/transpilation.cpp [hunk 4]
```diff
@@ -344,7 +344,7 @@ class TranspilationConsumer : public clang::ASTConsumer {
 
         auto traversal = std::make_unique<PreorderNlrTraversal>(_stage);
 
-        auto& input = _stage.getSession().getInput();
+        // auto& input = _stage.getSession().getInput();
```
Comment (IuriiKobein) [inline lib/pipeline/stages/transpiler/transpilation.cpp:347]: remove comment

File: lib/core/transpiler_session/transpiler_session.h (hunk 1 above)
Comment (pavlo-hilei) [reply lib/core/transpiler_session/transpiler_session.h]: Done

File: lib/core/transpiler_session/transpiler_session.h (hunk 2 above)
Comment (pavlo-hilei) [reply lib/core/transpiler_session/transpiler_session.h]: Done

File: lib/core/transpiler_session/transpiler_session.h (hunk 3 above)
Comment (pavlo-hilei) [reply lib/core/transpiler_session/transpiler_session.h]: Done

File: lib/core/transpiler_session/transpiler_session.h [hunk 5]
```diff
@@ -25,6 +25,11 @@ class SessionStage;
 using SharedTranspilerSession = std::shared_ptr<TranspilerSession>;
 
 struct TranspilerSession {
+    struct StagedFile {
```
[the remark this answers is not in the document]
Comment (pavlo-hilei) [reply lib/core/transpiler_session/transpiler_session.h]: Renamed

File: lib/core/transpiler_session/transpiler_session.h [hunk 6]
```diff
@@ -45,22 +50,28 @@ struct TranspilerSession {
...
+        _stagedFile.headers = _output.normalized.headers;
     }
 
    private:
     // TODO add methods for user input/output
-    UserInput _input;
+    const UserInput _input;
     UserOutput _output;
 
+    StagedFiles _stagedFile;
```
Comment (IuriiKobein) [inline lib/core/transpiler_session/transpiler_session.h:73]: let be consistent - _stagedFiles :)

Comment (pavlo-hilei) [reply lib/core/transpiler_session/transpiler_session.h:73]: I feel like chatGPT

Comment (pavlo-hilei) [reply lib/core/transpiler_session/transpiler_session.h]: Renamed

**Pull request merged**